### PR TITLE
[IMP] core: delay translation

### PR DESCRIPTION
**Pull request by HydrionBurst**
After writing a model terms translated field in one language, keep old translations for other languages until any user confirms those new translations from the website TRANSLATE page or the translation dialog.

DB:
Jsonb data structure of model_terms translated fields' columns
```sql
'{
	"en_US": "<div>Apple</div>"
	"fr_FR": "<div>Pomme</div><div>Banane</div>"
	"_fr_FR": "<div>Pomme</div>"
}'::jsonb
```

logic constraint
"column" IS NULL OR "column"->>'en_US' IS NOT NULL

`"column"->>'lang'`
1. stores last confirmed value
2. logically fallbacks to the following when read `lang`
```sql
    COALESCE(
        "column"->>'lang',
        "column"->>'en_US'
    )
```

`"column"->>'_lang'`
1. stores translations and the last written html/xml structure
2. shares the same html/xml structure with other "column"->>'_lang'
3. logically fallbacks to the following when read `lang` with context `check_translations`/`edit_translations`
```sql
    COALESCE(
        "column"->>'_lang',
        "column"->>'lang',
        "column"->>'_en_US',
        "column"->>'en_US'
    )
```

Data flow for column value, cache value and record value
(make sure the window is wide enough to see the graph)
contexts:
`check_translations`: used to read _lang value and get current translation mappings
`edit_translations`: used to read _lang value and wrap each term to let web client identify them

```
                             +- - - - - - - - - - - - - - - - - - - - - - - - - - - -+
                             ' Record(str):                                          '
                             '                                                       '
                             ' +---------------------------------------------------+ '               +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
                             ' | "<div><span data-oe-model='model'                 | '               {            read fr_FR             {
                             ' | data-oe-id='id' ...>French</span></div>"          | ' <------------ } context.get('edit_translations')  } <+
                             ' +---------------------------------------------------+ '               +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+  |
                             ' +---------------------------------------------------+ '               +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+  |
                             ' | "<div>French</div>"                               | '               {            read fr_FR             {  |
                             ' |                                                   | ' <------------ } context.get('check_translations') } <+
                             ' +---------------------------------------------------+ '               +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+  |
+~~~~~~~~~~~~+               ' +---------------------------------------------------+ '                                                      |
{ read fr_FR { ------------> ' | "<a>French</a>"                                   | '                                                      |
+~~~~~~~~~~~~+               ' +---------------------------------------------------+ '                                                      |
  ^                          '                                                       '                                                      |
  |                          +- - - - - - - - - - - - - - - - - - - - - - - - - - - -+                                                      |
  |                                                                                                                                         |
  |                                                                                                                                         |
  |                                                                                                                                         |
  |                          +- - - - - - - - - - - - - - - - - - - - - - - - - - - -+                                                      |
  |                          ' Cache(dict):                                          '                                                      |
  |                          '                                                       '                                                      |
  |                          ' +---------------------------------------------------+ '                                                      |
  |                          ' |                                                   | ' -----------------------------------------------------+
  |                          ' | "_fr_FR": "<div>French</div>"                     | '
  |                          ' |                                                   | ' <----------------------------------------------------+
  |                          ' +---------------------------------------------------+ '                                                      |
  |                          ' +---------------------------------------------------+ '               +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
  +------------------------- ' |                                                   | '               {              fetch fr_FR               {
                             ' | "fr_FR": "<a>French</a>"                          | '               }    context.get('edit_translations')    }
  +------------------------> ' |                                                   | '               {  or context.get('check_translations')  {
  |                          ' +---------------------------------------------------+ '               +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
+~~~~~~~~~~~~~~~~~+          '                                                       '                                                      ^
{   fetch fr_FR   {          +- - - - - - - - - - - - - - - - - - - - - - - - - - - -+                              COALESCE(               |
+~~~~~~~~~~~~~~~~~+                                                                                                     "c"->>'_fr_FR',     |
  ^                                                                                                                     "c"->>'fr_FR',      |
  | COALESCE(                                                                                                           "c"->>'_en_US',     |
  |     "c"->>'fr_FR',       +- - - - - - - - - - - - - - - - - - - - - - - - - - - -+                                  "c"->>'en_US'       |
  |     "c"->>'en_US'        ' Database(jsonb):                                      '                              )                       |
  | )                        '                                                       '                                                      |
  |                          ' +---------------------------------------------------+ '                                                      |
  |                          ' | "_fr_FR": "<div>French</div>"                     | ' -----------------------------------------------------+
  |                          ' +---------------------------------------------------+ '                                                      |
  |                          ' +---------------------------------------------------+ '                                                      |
  +------------------------- ' | "fr_FR": "<a>French</a>"                          | ' -----------------------------------------------------+
  |                          ' +---------------------------------------------------+ '                                                      |
  |                          ' +---------------------------------------------------+ '                                                      |
  |                          ' | "_en_US": "<div>English</div>"                    | ' -----------------------------------------------------+
  |                          ' +---------------------------------------------------+ '                                                      |
  |                          ' +---------------------------------------------------+ '                                                      |
  +------------------------- ' | "en_US": "<div>English1</div><div>English2</div>" | ' -----------------------------------------------------+
                             ' +---------------------------------------------------+ '
                             '                                                       '
                             +- - - - - - - - - - - - - - - - - - - - - - - - - - - -+
```

task: 3043340

### CODE REVIEW

**Review by rco-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Comment by rco-odoo** [comment]
@robodoo override=ci/security

**Comment by rdeodoo** [comment]
cc @pierre-pde FYI

**Comment by pierre-pde** [comment]
@HydrionBurst , thanks for this PR.
I found a bug:
On website, the first time you click on translate, every text is highlighted in green (meaning it's translated).
It should be highlighted in yellow (meaning it's not translated).
If you save & click on translate again, the colors are correct.
Video : https://bit.ly/498Qh28

cc @rdeodoo

**Comment by d-fence** [comment]
robodoo r-
I'm sorry to say that it [breaks the staging](https://runbot.odoo.com/runbot/build/52904060) when building with enterprise
Here is a [custom build](https://runbot.odoo.com/runbot/build/52906290) on you branch to reproduce the issue.

**Comment by HydrionBurst** [comment]
> @HydrionBurst , thanks for this PR. I found a bug: On website, the first time you click on translate, every text is highlighted in green (meaning it's translated). It should be highlighted in yellow (meaning it's not translated). If you save & click on translate again, the colors are correct. Video : https://bit.ly/498Qh28
> 
> cc @rdeodoo

Currently, it is a limitation, because I mark all terms as `o_dirty` for a delay translated template( these terms still have the attribute "to_translate").
And `dirty` terms are displayed as 'translated' when you open the website. 
So, when you save, all these terms will be sent to the server to update translations even if you don't modify anything.
I will try to find a workaround for the UI after freeze

File: odoo/api.py [hunk 1]
```diff
@@ -1064,14 +1066,20 @@ def insert_missing(self, records, field, values):
...
                     else:
-                        val_all_en = dict.fromkeys(langs, val['en_US'])
-                        field_cache[id_] = {**val_all_en, **val}
+                        if _langs:  # fallback missing _lang to lang if exists
+                            val.update({f'_{k}': v for k, v in val.items() if k in langs and f'_{k}' not in val})
+                        field_cache[id_] = {
+                            **dict.fromkeys(langs, val['en_US']),  # fallback missing lang to en_US
+                            **dict.fromkeys(_langs, val.get('_en_US')),  # fallback missing _lang to _en_US
+                            **val
+                        }
```
Comment (HydrionBurst) [inline odoo/api.py:1080]: @rco-odoo 
This blocked the runbot
When `prefetch_langs` while `check_translations` I forgot to fallback the cahce for all _langs, which cause a MissingError in the runbot test

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by HydrionBurst** [comment]
BTW The `to translate` bug is fixed in the latest code
cc @pierre-pde

**Pull request closed without merging**